crates/aptos-crypto/src/p256_ecdsa/p256_keys.rs

```diff
@@ -0,0 +1,275 @@
+        }
+    }
+
+    /// Private function aimed at minimizing code duplication between sign
+    /// methods of the SigningKey implementation. This should remain private.
+    /// This function uses the `RustCrypto` P256 signing library, which uses,
+    /// as of version 0.13.2, SHA2-256 as its hashing algorithm
+    fn sign_arbitrary_message(&self, message: &[u8]) -> P256Signature {
+        let secret_key = &self.0;
+        let sig = P256Signature(secret_key.sign(message.as_ref()));
+        P256Signature::make_canonical(&sig)
+    }
+}
+
+impl P256PublicKey {
+    /// Serialize a P256PublicKey. Uses the SEC1 serialization format.
+    pub fn to_bytes(&self) -> [u8; P256_PUBLIC_KEY_LENGTH] {
+        // The RustCrypto P256 `to_sec1_bytes` call here should never return an array of the wrong length and cause a panic
+        (*self.0.to_sec1_bytes()).try_into().unwrap()
+    }
+
+    /// Deserialize a P256PublicKey, checking expected key size
+    /// and that it is a valid curve point.
+    /// Uses the SEC1 serialization format.
+    pub(crate) fn from_bytes_unchecked(
+        bytes: &[u8],
+    ) -> std::result::Result<P256PublicKey, CryptoMaterialError> {
+        match p256::ecdsa::VerifyingKey::from_sec1_bytes(bytes) {
+            Ok(p256_public_key) => Ok(P256PublicKey(p256_public_key)),
+            Err(_) => Err(CryptoMaterialError::DeserializationError),
+        }
+    }
+}
+
+///////////////////////
+// PrivateKey Traits //
+///////////////////////
+
+impl PrivateKey for P256PrivateKey {
+    type PublicKeyMaterial = P256PublicKey;
+}
+
+impl SigningKey for P256PrivateKey {
+    type SignatureMaterial = P256Signature;
+    type VerifyingKeyMaterial = P256PublicKey;
+
+    fn sign<T: CryptoHash + Serialize>(
+        &self,
+        message: &T,
+    ) -> Result<P256Signature, CryptoMaterialError> {
+        Ok(P256PrivateKey::sign_arbitrary_message(
+            self,
+            signing_message(message)?.as_ref(),
+        ))
+    }
+
+    #[cfg(any(test, feature = "fuzzing"))]
+    fn sign_arbitrary_message(&self, message: &[u8]) -> P256Signature {
+        P256PrivateKey::sign_arbitrary_message(self, message)
+    }
+}
+
+impl Uniform for P256PrivateKey {
+    // Returns a random field element as a private key indistinguishable from uniformly random.
+    // Uses a hack to get around the incompatability of the `aptos-crypto` RngCore trait and the
+    // `RustCrypto` RngCore trait
+    fn generate<R>(rng: &mut R) -> Self
+    where
+        R: ::rand::RngCore + ::rand::CryptoRng + ::rand_core::CryptoRng + ::rand_core::RngCore,
+    {
+        let mut bytes = [0u8; P256_PRIVATE_KEY_LENGTH * 2];
+        rng.fill_bytes(&mut bytes);
+        let bignum = BigUint::from_bytes_be(&bytes[..]);
+        let order = BigUint::from_bytes_be(&ORDER);
+        let remainder = bignum.mod_floor(&order);
+        P256PrivateKey(p256::ecdsa::SigningKey::from_slice(&remainder.to_bytes_le()).unwrap())
```

Comment:
Will this panic if the `bytes` do not encode a valid scalar in that field?

Reply:
Also, why not call `P256PrivateKey::from_bytes_unchecked` here?

Reply:
You mean if `remainder` doesn't encode a valid scalar in the field? How would that happen when remainder is the result of `bytes` being taken mod the scalar field order?

Reply:
Changed to `from_bytes_unchecked`

Reply:
Check my logic on deciding `to_bytes_be` is correct here - I defined a constant `ORDER_TOO_MUCH` that is identical to `ORDER`, except the least significant (rightmost) byte is `0xFF`. I then convert it to a `BigUint` using `from_bytes_be` (this is correct from the tests I did on `ORDER`). Then I convert the BigUint into bytes again using `to_bytes_(le|be)`, and call `P256PrivateKey::from_bytes_unchecked` on the result. `_be` panics most of the p256 tests, which means it is the right option, as it correctly results in a value which is greater than the field order

Reply:
Added a unit test that makes this clearer. See https://github.com/aptos-labs/aptos-core/pull/9594/commits/a1506e2dfccaab44c7e1930c1e11b336d7f3e767